src/rgw/rgw_data_sync.h

```diff
@@ -347,19 +348,52 @@ void pretty_print(const RGWDataSyncEnv* env, T&& ...t) {
   }
 }
 
+/// \brief Adjust concurrency based on latency
+///
+/// Keep a running average of operation latency and scale concurrency
+/// down when latency rises.
+class LatencyConcurrencyControl : public LatencyMonitor {
+public:
+  CephContext* cct;
+
+  LatencyConcurrencyControl(CephContext* cct)
+    : cct(cct)  {}
+
+  /// \brief Lower concurrency when latency rises
+  ///
+  /// Since we have multiple spawn windows (data sync overall and
+  /// bucket), accept a number of concurrent operations to spawn and,
+  /// if latency is high, cut it in half. If latency is really high,
+  /// cut it to 1.
+  int64_t adj_concurrency(int64_t concurrency) {
+    using namespace std::literals;
+    auto threshold = (cct->_conf->rgw_sync_lease_period * 1s) / 12;
```

Comment:
@cbodley I was thinking about this, and I think making this scale with the least time might be the wrong idea.

If we have a twenty minute lease time then ten or twenty seconds for the lock action to complete is still pretty egregious and suggests the we're overloading the OSD, and with it scaling this way we'd end up only throttling back with hundred second average latency.

Reply:
(cc @mattbenjamin)

considering a deployment with one radosgw per zone where we could omit the leases entirely, this latency shouldn't matter at all - it could be 10 minutes or an hour, and sync would chug along (albeit slowly) without any errors

it wasn't until the `RGWContinuousLeaseCR::is_locked()` change from https://github.com/ceph/ceph/pull/47728 that this latency turned into actual sync errors. because this lease timer is the only thing that's sensitive to latency, i suggested that the throttling logic should aim to keep that latency in a range where we can actually make progress

however, lacking a better model for flow control or throttling here, maybe we should just revert that `is_locked()` part of https://github.com/ceph/ceph/pull/47728 - as-is, that seems to be making sync *less* reliable because it's shown that we have no control over this latency in our testing

i'm open to exploring band-aids here just to ship something, but this is complicated and it's hard for me to tell whether we're making things better or worse. even with your super-conservative throttling, we were still seeing extreme latencies here, right? that suggests the load is coming from somewhere else like ingest or /admin/log requests

Reply:
@cbodley I guess let's review today and go with your and @adamemerson 's intuition on this for 5.3, at least